### PR TITLE
[CPU] Dynamism support for RegionYolo node

### PR DESCRIPTION
**Pull request by eshoguli**
### Details:
 - *Extend the existing RegionYolo node with dynamic shapes support*

### Tickets:
 - *70637*

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@mandrono please review

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/region_yolo.cpp [hunk 1]
```diff
@@ -9,23 +9,65 @@
...
+        {{1, 75, 52, 52}, {1, 75, 32, 32}}
+    },
+    {
+        // input model dynamic shapes with interval limits
+        {1, {75, 75}, {32, 52}, {32, 52}},
+        // input tensor shapes
+        {{1, 75, 52, 52}, {1, 75, 32, 32}}
+    },
+    // static shapes
+    {{1, 75, 52, 52}, {{1, 75, 52, 52}}},
```
Comment (a-sidorova) [inline inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/region_yolo.cpp:45]: Minor: I think you can create new `std:;vector<ov::Shape>` with static shape and then create new separating instantiate for static shapes. I suggest to separate tests with static and dynamic shapes. What do you think?

File: inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/region_yolo.hpp [hunk 2]
```diff
@@ -27,12 +28,15 @@ using regionYoloParamsTuple = std::tuple<
         std::string>;                   // Device name
 
 class RegionYoloLayerTest : public testing::WithParamInterface<regionYoloParamsTuple>,
-                            virtual public LayerTestsUtils::LayerTestsCommon {
+                            virtual public ov::test::SubgraphBaseTest {
 public:
+    RegionYoloLayerTest() {
+        std::cout << "RegionYoloLayerTest::RegionYoloLayerTest" << std::endl;
```
Comment (a-sidorova) [inline inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/region_yolo.hpp:34]: For what it is?

File: inference-engine/tests/functional/shared_test_classes/src/base/ov_subgraph.cpp [hunk 3]
```diff
@@ -26,6 +26,11 @@
 namespace ov {
 namespace test {
 
+std::ostream& operator <<(std::ostream& os, const InputShape& inputShape) {
+    os << inputShape.first << "_" << CommonTestUtils::vec2str(inputShape.second);
```
Comment (a-sidorova) [inline inference-engine/tests/functional/shared_test_classes/src/base/ov_subgraph.cpp:30]: I think it's incorrect. Because `InputShape` is a pair when second part is vector of target shapes. Why do you use only target shapes? Where is dynamic shape?

File: inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp [hunk 4]
```diff
@@ -3,11 +3,13 @@
...
 
 #include "shared_test_classes/single_layer/region_yolo.hpp"
+#include "shared_test_classes/base/ov_subgraph.hpp"
 
 namespace LayerTestsDefinitions {
+using namespace ov::test;
 
 std::string RegionYoloLayerTest::getTestCaseName(const testing::TestParamInfo<regionYoloParamsTuple> &obj) {
-    ngraph::Shape inputShape;
+    ov::test::InputShape inputShape;
```
Comment (a-sidorova) [inline inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp:12]: First you use `using namespace ov::test` but then you use `ov::test::InputShape`, it's strange...

File: inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp [hunk 5]
```diff
@@ -42,8 +45,11 @@ void RegionYoloLayerTest::SetUp() {
     int end_axis;
     InferenceEngine::Precision netPrecision;
     std::tie(inputShape, classes, coords, num_regions, do_softmax, mask, start_axis, end_axis, netPrecision, targetDevice) = this->GetParam();
+
+    init_input_shapes({ inputShape });
+
     auto ngPrc = FuncTestUtils::PrecisionUtils::convertIE2nGraphPrc(netPrecision);
-    auto param = std::make_shared<ngraph::op::Parameter>(ngPrc, inputShape);
+    auto param = std::make_shared<ngraph::op::Parameter>(ngPrc, inputShape.first);
```
Comment (a-sidorova) [inline inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp:52]: You can use builder for parameter with dynamic shape. [Example](https://github.com/openvinotoolkit/openvino/blob/021477055387dadcd001021138b116b7880c3469/inference-engine/tests/functional/plugin/cpu/single_layer_tests/normalize.cpp#L77)

**Review by a-sidorova — CHANGES_REQUESTED** [review]

File: inference-engine/tests/functional/inference_engine/serialization/single_layer/region_yolo.cpp [hunk 6]
```diff
@@ -11,14 +11,14 @@ using namespace LayerTestsDefinitions;
...
-        Serialize();
+        serialize();
     }
 
 
-    const std::vector<ngraph::Shape> inShapes_v3 = {
-        {1, 255, 52, 52},
-        {1, 255, 26, 26},
-        {1, 255, 13, 13}
+    const std::vector<ov::test::InputShape> inShapes_v3 = {
```
[the remark this answers is not in the document]
Comment (eshoguli) [reply inference-engine/tests/functional/inference_engine/serialization/single_layer/region_yolo.cpp:18]: done, thanks

File: inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/region_yolo.hpp (hunk 2 above)
Comment (eshoguli) [reply inference-engine/tests/functional/shared_test_classes/include/shared_test_classes/single_layer/region_yolo.hpp:34]: done, thanks

File: inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp (hunk 4 above)
Comment (eshoguli) [reply inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp:12]: type full name has been used before `using namespace ov::test` was added. Fixed, thanks

File: inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp (hunk 5 above)
Comment (eshoguli) [reply inference-engine/tests/functional/shared_test_classes/src/single_layer/region_yolo.cpp:52]: updated, thanks. Some notes: 1) original source code used is this approach 2) to test support engineers: bad practice to have not really neccessary (`inputDynamicShapes`) base type members

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/region_yolo.cpp (hunk 1 above)
Comment (eshoguli) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/region_yolo.cpp:45]: there are not many changes here. I would like to keep as is

File: inference-engine/tests/functional/shared_test_classes/src/base/ov_subgraph.cpp (hunk 3 above)
Comment (eshoguli) [reply inference-engine/tests/functional/shared_test_classes/src/base/ov_subgraph.cpp]: fixed, the commit was cherry-picked. Has been discussed with @maxnick:   https://github.com/openvinotoolkit/openvino/pull/8387#discussion_r751342702

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_region_yolo_node.h [hunk 7]
```diff
@@ -49,6 +49,10 @@ class MKLDNNRegionYoloNode : public MKLDNNNode {
 
     static bool isSupportedOperation(const std::shared_ptr<const ngraph::Node>& op, std::string& errorMessage) noexcept;
 
+protected:
+    void prepareParams() override;
```
Comment (a-sidorova) [inline inference-engine/src/mkldnn_plugin/nodes/mkldnn_region_yolo_node.h:53]: You should just override `needPrepareParams` to return false. Because you don't need prepare params for each new shapes. So you may not override `prepareParams()`. `needPrepareParams` is enough

Moreover you should update last input dims in the end of `createPrimitive` if shapes are defined. [Example](https://github.com/v-Golubev/openvino/blob/1e19e7f74f6d39fc00c472d2ce3ddca791cecc17/inference-engine/src/mkldnn_plugin/nodes/mkldnn_cum_sum_node.cpp#L291)

**Review by a-sidorova — CHANGES_REQUESTED** [review]

**Review by a-sidorova — APPROVED** [review]

**Comment by eshoguli** [comment]
@jane-intel & @ilyachur, can you, please, review `RegionYolo` nGraph operation update: ngraph/core/src/op/region_yolo.cpp (https://github.com/openvinotoolkit/openvino/pull/8542/files#diff-5eb4be30d8c1e52c4faa67fae323f45aaf08dd0e769b53c49a28185fcea7de33) ? 
thanks!

File: inference-engine/src/mkldnn_plugin/nodes/mkldnn_region_yolo_node.h (hunk 7 above)
Comment (eshoguli) [reply inference-engine/src/mkldnn_plugin/nodes/mkldnn_region_yolo_node.h]: done, thanks

File: inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/region_yolo.cpp (hunk 1 above)
Comment (eshoguli) [reply inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/region_yolo.cpp]: fixed, thanks

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@eshoguli please make CI green

**Comment by dmitry-gorokhov** [comment]
@jane-intel @ilyachur any comments from your side? I am going to merge the PR.

File: ngraph/core/src/op/region_yolo.cpp [hunk 8]
```diff
@@ -69,15 +70,19 @@ void op::RegionYolo::validate_and_infer_types() {
                 output_shape.push_back(input_shape[i]);
             }
             for (int64_t i = m_axis; i < end_axis + 1; i++) {
-                flat_dim *= input_shape[i];
+                if (input_shape[i].is_dynamic()) {
```
Comment (jane-intel) [inline ngraph/core/src/op/region_yolo.cpp:73]: I have a better way to formulate this piece of code. Will do it in the separate pr.

**Review by jane-intel — APPROVED** [review]

File: ngraph/core/src/op/region_yolo.cpp (hunk 8 above)
Comment (eshoguli) [reply ngraph/core/src/op/region_yolo.cpp]: can you provide link to PR?

**Review by dmitry-gorokhov — APPROVED** [review]
